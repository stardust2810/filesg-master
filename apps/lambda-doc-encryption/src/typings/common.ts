import { FileExtension, MimeType } from 'file-type';

export type AgencyPasswordFileInfo = {
  filePath: string;
  absoluteFilePath: string;
  ext: FileExtension | undefined;
  mime: MimeType | undefined;
  size: number;
};
