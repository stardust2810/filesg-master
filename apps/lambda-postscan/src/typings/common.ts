// =============================================================================
// Typing
// =============================================================================
export interface LambdaSnsEvent {
  Records: LambdaSnsRecord[];
}

interface LambdaSnsRecord {
  EventVersion: string;
  EventSubscriptionArn: string;
  EventSource: string;
  Sns: {
    SignatureVersion: string;
    Timestamp: string;
    Signature: string;
    SigningCertUrl: string;
    MessageId: string;
    Message: string;
    MessageAttributes: {
      Test: {
        Type: string;
        Value: string;
      };
      TestBinary: {
        Type: string;
        Value: string;
      };
    };
    Type: string;
    UnsubscribeUrl: string;
    TopicArn: string;
    Subject: string;
  };
}

export type ScanResultSnsMessage = ScanResultSuccessSnsMessage | ScanResultFailureSnsMessage;

interface ScanResultBaseMessage {
  timestamp: number;
  sqs_message_id: string;
  xamz_request_id: string;
  file_url: string;
  scanner_status_message: string;
}

interface ScanResultSuccessSnsMessage extends ScanResultBaseMessage {
  scanner_status: 0;
  scanning_result: {
    TotalBytesOfFile: number;
    Findings: ScanResultFindings[];
    Error: string;
    Codes: number[];
  };
}

interface ScanResultFailureSnsMessage extends ScanResultBaseMessage {
  scanner_status: -1 | -2 | -3 | -4;
  scanning_result: {
    Error: string;
  };
}

interface ScanResultFindings {
  malware: string;
  type: string;
}

// =============================================================================
// Constants
// =============================================================================
// Trend Micro Cloud One FSS
export const CODE_SKIP_MULTIPLE = 100;
export const CODE_MISC = 199;
export const FSS_TAG_PREFIX = 'fss-';

// =============================================================================
// Enums
// =============================================================================
export enum SCAN_RESULT_MESSAGE {
  'incomplete scan due to multiple reasons' = 100,
  'incomplete archive file extraction due to file too large',
  'incomplete archive file extraction due to too many files in archive',
  'incomplete archive file extraction due to too many archive layers',
  'incomplete archive file extraction due to compression ratio exceeds limit',
  'incomplete archive file extraction due to unsupported compression method',
  'incomplete archive file extraction due to corrupted compression file',
  'incomplete archive file extraction due to archive file encryption',
  'incomplete scan due to Microsoft Office file encryption',
  'incomplete scan due to miscellaneous reason. Provide the fss-scan-detail-code tag value to Trend Micro support' = 199,
}

export enum SCAN_RESULT_TAG {
  NO_ISSUES_FOUND = 'no issues found',
  MALICIOUS = 'malicious',
}
