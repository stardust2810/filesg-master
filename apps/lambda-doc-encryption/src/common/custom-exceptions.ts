import { FileSGBaseException } from '@filesg/backend-common';
import { COMPONENT_ERROR_CODE } from '@filesg/common';

import { AgencyPasswordFileInfo } from '../typings/common';

export class FileDownloadException extends FileSGBaseException {
  constructor(message: string, componentErrorCode: COMPONENT_ERROR_CODE, key: string) {
    super(`[FileDownloadException] Error when downloading file with key ${key}: ${message}`, componentErrorCode);
  }
}

export class UnsupportedFileTypeException extends FileSGBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, contentType: string) {
    super(`[UnsupportedFileTypeException] File type of ${contentType} is not supported`, componentErrorCode);
  }
}

export class GetFileSizeException extends FileSGBaseException {
  constructor(message: string, componentErrorCode: COMPONENT_ERROR_CODE, key: string) {
    super(`[GetFileSizeException] Error when getting size of object with key ${key}: ${message}`, componentErrorCode);
  }
}

export class UnsupportedFileTypeForEncryptionException extends FileSGBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, mainFilename: string, fileinfo: AgencyPasswordFileInfo[]) {
    super(
      `[UnsupportedFileTypeForEncryptionException] File type is not supported for password protection. ${JSON.stringify({
        [mainFilename]: fileinfo.map(({ filePath, mime }) => ({ filePath, mime })),
      })}`,
      componentErrorCode,
    );
  }
}

export class MissingFileException extends FileSGBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, mainFilename: string, missingFiles: string[]) {
    super(`[MissingFileException] Missing files: ${JSON.stringify({ [mainFilename]: missingFiles })}`, componentErrorCode);
  }
}
