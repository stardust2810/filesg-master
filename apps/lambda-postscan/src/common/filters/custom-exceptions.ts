import { HttpException, HttpStatus, InternalServerErrorException } from '@nestjs/common';

export class AWSException extends InternalServerErrorException {
  constructor(err: string) {
    super(`[AWSException] Error from AWS: ${err}`);
  }
}

export class InputValidationException extends HttpException {
  constructor(error: any) {
    super(
      {
        message: '[InputValidationException] Input data validation failed',
        error,
      },
      HttpStatus.BAD_REQUEST,
    );
  }
}
