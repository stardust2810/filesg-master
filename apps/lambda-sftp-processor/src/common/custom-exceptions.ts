import { FileSGBaseException } from '@filesg/backend-common';
import { COMPONENT_ERROR_CODE } from '@filesg/common';
import { ValidationError } from 'class-validator';

import { SLIFT_ENCRYPTED_FILE_EXT } from '../const';

export class SftpProcessorBaseException extends FileSGBaseException {
  constructor(
    message: string,
    componentErrorCode: COMPONENT_ERROR_CODE,
    public readonly toRetryProcessing: boolean,
    errorData?: string | string[] | Record<string, any>,
  ) {
    super(message, componentErrorCode, errorData);
  }
}

// =============================================================================
// Retry-able exception
// =============================================================================
export class ProcessMessageErrorException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, additionalMessage: string, internalLog?: string) {
    super(`[ProcessMessageErrorException] Error when processing SFTP SQS message: ${additionalMessage}`, componentErrorCode, true);
    this.internalLog = internalLog;
  }
}

export class FileDownloadErrorException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, internalLog?: string) {
    super(`[FileDownloadErrorException] Error when downloading sftp file`, componentErrorCode, true);
    this.internalLog = internalLog;
  }
}

export class SliftDecryptException extends SftpProcessorBaseException {
  constructor(path: string, err: Error) {
    super(
      `[SliftDecryptException] Failed to decrypt file ${path}. Due to the following error: ${err}`,
      COMPONENT_ERROR_CODE.SLIFT_SERVICE,
      true,
    );
  }
}

export class SliftCertGenerationException extends SftpProcessorBaseException {
  constructor(err: Error) {
    super(`[SliftCertGenerationException] Failed to create receiver private key.[Error]: ${err}`, COMPONENT_ERROR_CODE.SLIFT_SERVICE, true);
  }
}

export class SMGetSecretException extends SftpProcessorBaseException {
  constructor(key: string) {
    super(`[SMGetSecretException] Expecting the value for key: ${key} to be not empty`, COMPONENT_ERROR_CODE.SECRET_MANAGER_SERVICE, true);
  }
}

export class RequestTimeoutException extends SftpProcessorBaseException {
  constructor(sourceEndpoint: string, originErrorMsg: string) {
    super(
      `[RequestTimeoutException] Request timeout when calling ${sourceEndpoint} API: ${originErrorMsg}`,
      COMPONENT_ERROR_CODE.SFTP_PROCESSOR_SERVICE,
      true,
    );
  }
}

// =============================================================================
// Non retry-able exception
// =============================================================================
export class MissingSidecarFileException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, missingFiles: string[]) {
    super(`[MissingSidecarFileException] Missing sidecar files: ${JSON.stringify(missingFiles)}`, componentErrorCode, false, missingFiles);
  }
}

export class ParsingCsvException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, errorData: { filename: string; msg: string }[]) {
    super(`[ParsingCsvException] Error parsing csv files`, componentErrorCode, false, errorData);
  }
}

export class CsvValidationException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, validationErrors: ValidationError[]) {
    super(`[CsvValidationException] Error parsing csv files`, componentErrorCode, false, validationErrors);
  }
}

export class MissingFileException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, missingFiles: string[]) {
    super(`[MissingFileException] Missing files: ${JSON.stringify(missingFiles)}`, componentErrorCode, false, missingFiles);
  }
}

export class ExtraFileException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, extraFiles: string[]) {
    super(`[ExtraFileException] Extra files: ${JSON.stringify(extraFiles)}`, componentErrorCode, false, extraFiles);
  }
}

export class ExtraFileAgencyPasswordException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, extraFiles: string[]) {
    super(
      `[ExtraFileAgencyPasswordException] Extra files: ${JSON.stringify(extraFiles)} found inside agency password sidecar data.`,
      componentErrorCode,
      false,
      extraFiles,
    );
  }
}

export class InvalidFilePathFormatAgencyPasswordException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, invalidFiles: string[]) {
    super(
      `[InvalidFilePathFormatAgencyPasswordException] Invalid file path format. No file path found under the following files: ${JSON.stringify(
        invalidFiles,
      )}`,
      componentErrorCode,
      false,
      invalidFiles,
    );
  }
}

export class DuplicateEntryAgencyPasswordException extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE) {
    super(`[DuplicateEntryAgencyPasswordException] Agency password file contains duplicate entry.`, componentErrorCode, false);
  }
}

export class UnsupportedFileTypeException extends SftpProcessorBaseException {
  constructor(componentErroCode: COMPONENT_ERROR_CODE, unsupportedFiles: { name: string; detectedType: string | undefined }[]) {
    super(
      `[UnsupportedFileTypeException] Unsupported files: ${JSON.stringify(unsupportedFiles)}`,
      componentErroCode,
      false,
      unsupportedFiles,
    );
  }
}

export class UploadWorkingFileToS3Exception extends SftpProcessorBaseException {
  constructor(componentErrorCode: COMPONENT_ERROR_CODE, errorData: { key: string; msg: string }[]) {
    super(`[UploadWorkingFileToS3Exception] Error uploading working files to s3`, componentErrorCode, false, errorData);
  }
}

export class SliftMissingException extends SftpProcessorBaseException {
  constructor(sliftDir: string) {
    super(
      `[SliftMissingException] Unable to locate run shell script from SLIFT under ${sliftDir}`,
      COMPONENT_ERROR_CODE.SLIFT_SERVICE,
      false,
    );
  }
}

export class SliftFileTypeException extends SftpProcessorBaseException {
  constructor() {
    super(
      `[SliftFileTypeException] File is not a slift encrypyted file as the extenstion is not ${SLIFT_ENCRYPTED_FILE_EXT}`,
      COMPONENT_ERROR_CODE.SLIFT_SERVICE,
      false,
    );
  }
}

export class CreateTransactionException extends SftpProcessorBaseException {
  constructor(originErrorMsg: string, actualErrorRes: Record<string, unknown>) {
    super(
      `[CreateTransactionException] Error when calling service-core CreateTransaction API: ${originErrorMsg}`,
      COMPONENT_ERROR_CODE.SFTP_PROCESSOR_SERVICE,
      false,
      actualErrorRes,
    );
  }
}

export class UploadTransactionFilesException extends SftpProcessorBaseException {
  constructor(originErrorMsg: string, actualErrorRes: Record<string, unknown>) {
    super(
      `[UploadTransactionFilesException] Error when calling service-transfer UploadFile API: ${originErrorMsg}`,
      COMPONENT_ERROR_CODE.SFTP_PROCESSOR_SERVICE,
      false,
      actualErrorRes,
    );
  }
}
