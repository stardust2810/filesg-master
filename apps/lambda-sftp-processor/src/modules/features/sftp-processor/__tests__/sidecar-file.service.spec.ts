import { CsvHelper, FsHelper } from '@filesg/backend-common';
import { Test, TestingModule } from '@nestjs/testing';

import { CsvValidationException, MissingSidecarFileException } from '../../../../common/custom-exceptions';
import {
  FILE_SIDECAR_FILENAME,
  NOTIFICATIONS_SIDECAR_FILENAME,
  RECIPIENT_SIDECAR_FILENAME,
  TRANSACTION_SIDECAR_FILENAME,
} from '../../../../const';
import {
  mockDir,
  mockSidecareNotificationsRecords,
  mockSidecarFileRecords,
  mockSidecarRecipientRecords,
  mockSidecarTransactionRecord,
} from '../__mocks__/sidecar-file.service.mock';
import { SidecarFileService } from '../sidecar-file.service';

describe('SidecarFileService', () => {
  let service: SidecarFileService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [SidecarFileService],
    }).compile();

    service = module.get<SidecarFileService>(SidecarFileService);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('checkSidecarFilesExistsOrThrow', () => {
    it('should not throw exception when there is no missing files', async () => {
      jest.spyOn(FsHelper, 'checkForMissingFilesInDir').mockResolvedValueOnce([]);

      await expect(service.checkSidecarFilesExistsOrThrow(mockDir)).resolves.not.toThrow();
      expect(FsHelper.checkForMissingFilesInDir).toBeCalledWith(mockDir, [
        TRANSACTION_SIDECAR_FILENAME,
        RECIPIENT_SIDECAR_FILENAME,
        FILE_SIDECAR_FILENAME,
        NOTIFICATIONS_SIDECAR_FILENAME,
      ]);
    });

    it('MissingSidecarFileException should be thrown when there are missing sidecar files', async () => {
      jest.spyOn(FsHelper, 'checkForMissingFilesInDir').mockResolvedValueOnce([TRANSACTION_SIDECAR_FILENAME]);

      await expect(service.checkSidecarFilesExistsOrThrow(mockDir)).rejects.toThrowError(MissingSidecarFileException);
    });
  });

  describe('parseSidecarFiles', () => {
    it('should not throw exception when there is no validation error', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce(mockSidecarRecipientRecords)
        .mockResolvedValueOnce(mockSidecarFileRecords)
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      await expect(service.parseSidecarFiles(mockDir)).resolves.not.toThrow();
    });

    it('CsvValidationException should be thrown when there are more than 1 transaction record', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce([...mockSidecarTransactionRecord, ...mockSidecarTransactionRecord])
        .mockResolvedValueOnce(mockSidecarRecipientRecords)
        .mockResolvedValueOnce(mockSidecarFileRecords)
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      await expect(service.parseSidecarFiles(mockDir)).rejects.toThrow(CsvValidationException);
    });

    it('CsvValidationException should be thrown when mandatory fields are missing', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce([{}])
        .mockResolvedValueOnce([{}])
        .mockResolvedValueOnce([{}])
        .mockResolvedValueOnce([{}]);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when date fields are of wrong format', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce([{ ...mockSidecarRecipientRecords[0], dob: '01-01-1999' }])
        .mockResolvedValueOnce([{ ...mockSidecarFileRecords[0], expiry: '01-01-2999', deleteAt: '12-12-2999' }])
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when dob is missing but contact was provided', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce([{ ...mockSidecarRecipientRecords[0], dob: undefined }])
        .mockResolvedValueOnce([{ ...mockSidecarFileRecords[0] }])
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when contact is missing but dob was provided', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce([{ ...mockSidecarRecipientRecords[0], contact: undefined }])
        .mockResolvedValueOnce([{ ...mockSidecarFileRecords[0] }])
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when date fields does not fulfil requirements', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce([{ ...mockSidecarRecipientRecords[0], dob: '2999-01-01' }]) // future date
        .mockResolvedValueOnce([{ ...mockSidecarFileRecords[0], expiry: '1990-01-01', deleteAt: '1990-12-12' }]) // past date
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when expiry date is later than delete date', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce(mockSidecarRecipientRecords)
        .mockResolvedValueOnce([{ ...mockSidecarFileRecords[0], expiry: '2050-12-12', deleteAt: '2050-12-11' }]) // past date
        .mockResolvedValueOnce(mockSidecareNotificationsRecords);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when there is no notification channels', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce(mockSidecarRecipientRecords)
        .mockResolvedValueOnce(mockSidecarFileRecords) // past date
        .mockResolvedValueOnce([]);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when unknown notification channels is provided', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce(mockSidecarRecipientRecords)
        .mockResolvedValueOnce(mockSidecarFileRecords) // past date
        .mockResolvedValueOnce([{ ...mockSidecareNotificationsRecords[0], channel: 'UKNOWN' }]);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });

    it('CsvValidationException should be thrown when duplicate notification channels is provided', async () => {
      jest
        .spyOn(CsvHelper, 'convertCsvToJson')
        .mockResolvedValueOnce(mockSidecarTransactionRecord)
        .mockResolvedValueOnce(mockSidecarRecipientRecords)
        .mockResolvedValueOnce(mockSidecarFileRecords) // past date
        .mockResolvedValueOnce([mockSidecareNotificationsRecords[0], mockSidecareNotificationsRecords[0]]);

      try {
        await service.parseSidecarFiles(mockDir);
      } catch (error) {
        expect(error).toBeInstanceOf(CsvValidationException);

        expect((error as CsvValidationException).errorData).toMatchSnapshot();
      }

      expect.assertions(2);
    });
  });

  it('CsvValidationException should be thrown when file path is empty inside agency password sidecar file', async () => {
    jest
      .spyOn(CsvHelper, 'convertCsvToJson')
      .mockResolvedValueOnce(mockSidecarTransactionRecord)
      .mockResolvedValueOnce(mockSidecarRecipientRecords)
      .mockResolvedValueOnce(mockSidecarFileRecords)
      .mockResolvedValueOnce(mockSidecareNotificationsRecords)
      .mockResolvedValueOnce([{ filePath: '', password: 'P@ssw0rd' }]);

    try {
      await service.parseSidecarFiles(mockDir);
    } catch (error) {
      expect(error).toBeInstanceOf(CsvValidationException);

      expect((error as CsvValidationException).errorData).toMatchSnapshot();
    }

    expect.assertions(2);
  });

  it('CsvValidationException should be thrown when password is empty inside agency password sidecar file', async () => {
    jest
      .spyOn(CsvHelper, 'convertCsvToJson')
      .mockResolvedValueOnce(mockSidecarTransactionRecord)
      .mockResolvedValueOnce(mockSidecarRecipientRecords)
      .mockResolvedValueOnce(mockSidecarFileRecords)
      .mockResolvedValueOnce(mockSidecareNotificationsRecords)
      .mockResolvedValueOnce([{ filePath: 'IssuanceFile.zip/folder1/file.pdf', password: '' }]);

    try {
      await service.parseSidecarFiles(mockDir);
    } catch (error) {
      expect(error).toBeInstanceOf(CsvValidationException);

      expect((error as CsvValidationException).errorData).toMatchSnapshot();
    }

    expect.assertions(2);
  });
});
