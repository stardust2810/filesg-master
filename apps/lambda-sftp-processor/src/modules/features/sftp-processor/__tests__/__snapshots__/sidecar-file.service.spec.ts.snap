// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidecarFileService CsvValidationException should be thrown when file path is empty inside agency password sidecar file 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "filePath should not be empty",
            },
            "property": "filePath",
          },
        ],
        "property": "0",
      },
    ],
    "property": "agencyPassword",
  },
]
`;

exports[`SidecarFileService CsvValidationException should be thrown when password is empty inside agency password sidecar file 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "password should not be empty",
            },
            "property": "password",
          },
        ],
        "property": "0",
      },
    ],
    "property": "agencyPassword",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when contact is missing but dob was provided 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "EnhancedValidateIf": "dob must not be defined if contact is empty",
            },
            "property": "dob",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "EnhancedValidateIf": "contact must be a non-empty, valid string when dob is non-empty",
            },
            "property": "contact",
          },
        ],
        "property": "0",
      },
    ],
    "property": "recipients",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when date fields are of wrong format 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "EnhancedValidateIf": "dob must be a non-empty, valid, less than or equal to current date, and in the format of (yyyy-mm-dd) when contact is non-empty",
            },
            "property": "dob",
          },
        ],
        "property": "0",
      },
    ],
    "property": "recipients",
  },
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "IsValidFileSGDate": "Date input must be valid, larger than current date, and in the format of (yyyy-mm-dd)",
            },
            "property": "expiry",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "IsValidFileSGDate": "Date input must be valid, larger than current date, and in the format of (yyyy-mm-dd)",
            },
            "property": "deleteAt",
          },
        ],
        "property": "0",
      },
    ],
    "property": "files",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when date fields does not fulfil requirements 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "EnhancedValidateIf": "dob must be a non-empty, valid, less than or equal to current date, and in the format of (yyyy-mm-dd) when contact is non-empty",
            },
            "property": "dob",
          },
        ],
        "property": "0",
      },
    ],
    "property": "recipients",
  },
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "IsValidFileSGDate": "Date input must be valid, larger than current date, and in the format of (yyyy-mm-dd)",
            },
            "property": "expiry",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "IsValidFileSGDate": "Date input must be valid, larger than current date, and in the format of (yyyy-mm-dd)",
            },
            "property": "deleteAt",
          },
        ],
        "property": "0",
      },
    ],
    "property": "files",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when dob is missing but contact was provided 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "EnhancedValidateIf": "dob must be a non-empty, valid, less than or equal to current date, and in the format of (yyyy-mm-dd) when contact is non-empty",
            },
            "property": "dob",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "EnhancedValidateIf": "contact must not be defined if dob is empty",
            },
            "property": "contact",
          },
        ],
        "property": "0",
      },
    ],
    "property": "recipients",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when duplicate notification channels is provided 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "arrayUnique": "All notifications's elements must be unique",
    },
    "property": "notifications",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when expiry date is later than delete date 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "IsAfterDate": "Input must be after expiry date",
            },
            "property": "deleteAt",
          },
        ],
        "property": "0",
      },
    ],
    "property": "files",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when mandatory fields are missing 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "applicationType should not be empty",
              "isString": "applicationType must be a string",
            },
            "property": "applicationType",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "transactionName should not be empty",
              "isString": "transactionName must be a string",
            },
            "property": "transactionName",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isEnum": "transactionType must be one of the following values: upload, share, transfer, upload_share, upload_transfer, widget, revoke, expire, delete, recall",
              "isNotEmpty": "transactionType should not be empty",
            },
            "property": "transactionType",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isBoolean": "isAcknowledgementRequired must be a boolean value",
            },
            "property": "isAcknowledgementRequired",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "clientId should not be empty",
              "isString": "clientId must be a string",
            },
            "property": "clientId",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "clientSecret should not be empty",
              "isString": "clientSecret must be a string",
            },
            "property": "clientSecret",
          },
        ],
        "property": "0",
      },
    ],
    "property": "transactions",
  },
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "uin should not be empty",
              "isValidUin": "Input must be valid uin",
            },
            "property": "uin",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "fullName should not be empty",
              "isString": "fullName must be a string",
            },
            "property": "fullName",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isEmail": "email must be an email",
              "isNotEmpty": "email should not be empty",
            },
            "property": "email",
          },
        ],
        "property": "0",
      },
    ],
    "property": "recipients",
  },
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "name should not be empty",
              "isString": "name must be a string",
              "isValidFilename": "Input must be a valid filename",
            },
            "property": "name",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isHash": "checksum must be a hash of type sha256",
              "isNotEmpty": "checksum should not be empty",
              "isString": "checksum must be a string",
            },
            "property": "checksum",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isBoolean": "isPasswordEncryptionRequired must be a boolean value",
            },
            "property": "isPasswordEncryptionRequired",
          },
        ],
        "property": "0",
      },
    ],
    "property": "files",
  },
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isEnum": "channel must be one of the following values: EMAIL, SG_NOTIFY",
              "isNotEmpty": "channel should not be empty",
            },
            "property": "channel",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "templateId should not be empty",
              "isString": "templateId must be a string",
            },
            "property": "templateId",
          },
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotEmpty": "templateInput should not be empty",
              "isRecord": "The templateInput must be a Record<string, string> where both the key and value are strings and should not be empty.",
            },
            "property": "templateInput",
          },
        ],
        "property": "0",
      },
    ],
    "property": "notifications",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when there is no notification channels 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "arrayMinSize": "notifications must contain at least 1 elements",
    },
    "property": "notifications",
  },
]
`;

exports[`SidecarFileService parseSidecarFiles CsvValidationException should be thrown when unknown notification channels is provided 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isEnum": "channel must be one of the following values: EMAIL, SG_NOTIFY",
            },
            "property": "channel",
          },
        ],
        "property": "0",
      },
    ],
    "property": "notifications",
  },
]
`;
